### ISSUE TITLE
Improve operations scheduling in Actuator endpoints

### ISSUE THREAD

**Issue by bclozel** [spring-projects/spring-boot#17489]
After chatting with @bsideup and @smaldini , it seems we could improve the scheduling of reactive operations in Actuator. Currently, if the return type is not `Publisher` we're scheduling work on an elastic scheduler (creating/borrowing a thread per call).

Two examples of that in the current codebase:

* the prometheus endpoint returns a `String`, whereas it is not blocking
* the health indicator is caching the result, but it's still scheduling work on the elastic scheduler. We should try and short-circuit this o avoid creating new threads even though the value is cached

There are several ways to solve that:

1. wrap any "blocking type" that is known to be not blocking in a type that is supported by Spring Framework's `ReactiveAdapterRegistry` even if reactor is not present, like `CompletionStage`.
2. provide a marker annotation for tagging blocking / nonblocking operations. Current trends (r2dbc and project Loom) make that alternative less and less appealing.
3. use a solution provided by the Reactor team or the Spring Framework team

There is still a dedicated issue in Spring Framework that was triggered by the Actuator implementation for the reactive stack, see spring-projects/spring-framework#21184.

The reactor team is also thinking about applying bounds to the number of created threads (currently it is unbounded) and instead queuing tasks in an unbounded queue. With that, applications would be less likely to run into OOM memory issues.

**Comment by wilkinsona** [comment]
> the prometheus endpoint returns a String, whereas it is not blocking

It isn't non-blocking in 100% of cases. There are some `synchronized` blocks in the Prometheus classes that the endpoint uses so we'd need to be really careful that they're not in the code path that the endpoint exercises. For me, this shows the danger, particularly in third-party code, of assuming that code that looks like it may block is, in fact, non-blocking. It's quite hard to do in the first place and would need to be checked every time a dependency was updated.

> wrap any "blocking type" that is known to be not blocking in a type that is supported by Spring Framework's ReactiveAdapterRegistry even if reactor is not present, like CompletionStage.

I think it would be a shame if we required users to do this. I really like that third-party endpoints can be written today and using with both the Servlet and Reactive stacks without the third-party having to consider Reactive's requirements. Perhaps that's naive, though, and it's just not possible to shield users like this?

> The reactor team is also thinking about applying bounds to the number of created threads (currently it is unbounded) and instead queuing tasks in an unbounded queue. With that, applications would be less likely to run into OOM memory issues.

This feels like the ideal solution to me if it's possible. It avoids the risk of accidentally denoting a blocking endpoint as non-blocking while also, hopefully, avoiding runaway thread creation.
